### PR TITLE
REVEAL: Adding the title Templates to the content type YAML

### LINKED ISSUE #1563: REVEAL: Adding the title Templates to the content type YAML
# Core Tripal 4 Development Task

## Pull #1583

**This issue may be as simple as uncommenting some lines in the content types YAML, bringing back a dockblock comment in the Services/TripalContentTypes::createContentType() and adding testing to [the existing test for that service](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/tests/src/Functional/Services/TripalContentTypesTest.php). Of course manual testing will also be needed.**

## Current Issue Description

Originally, title defaults were created by the Tripal Content Type entity in [TripalEntityType->getDefaultTitleFormat()](https://github.com/tripal/tripal/blob/4.x/tripal/src/Entity/TripalEntityType.php#L457-L514). Then any module could implement hook_tripal_default_title_format() to add additional suggestions.

While this works, it is more difficult to change defaults then it needs to be with updates introduced in Drupal 8.

Instead PR #1413, added the title default per content type to the content type YAML file (they are still there but commented out currently) and then used that information to create the defaults.

**The following code links are in the original branch and show how it may have been implemented:**
The [TripalEntityType->getDefaultTitleFormat()](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/src/Entity/TripalEntityType.php#L468-L529) remains unchanged, and when getting the title format, it still checks to see if the title format has been set and only if it hasn't is getDefaultTitleFormat() called.

The difference comes in creating the content types in the first place:
- Since the title format is now included in the YAML, you will see in the service creating content types that it is [passed in through the details](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/src/Services/TripalContentTypes.php#L167).
- This means it gets passed into [TripalEntityType::create() here](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/src/Services/TripalContentTypes.php#L196).
- And then passed into the [parent::create() here](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/src/Entity/TripalEntityType.php#L198) which essentially sets TripalEntityType::title_format which means that if it is set in the YAML then [TripalEntityType->getDefaultTitleFormat()](https://github.com/tripal/tripal/blob/tv4g1-1393-ContentTypesYML/tripal/src/Entity/TripalEntityType.php#L468-L529) will not get called.

**Bringing this functionality back may be as simple as commenting out the title formats in the Yaml and adding back in the documentation line linked above!**

## Background

The background for this issue is a cautionary tale: Once upon a time a PR was made, #1413, with so many great changes for Tripal 4. This PR was focused on moving content type + field creation into YAML files for the good of all. This was a large task in and of itself but as such tasks have a habit of doing... it picked up many additional bug fixes and features as it grew. The developer was testing locally both manually and running the automated tests they developed just as they should but forgot to push out to github as often as recommended 😨 This caused the developer to miss a test failure in unrelated code within their beautiful PR 😱 As the PR had become an inter-related collection of fixes by the point it came up for review and the unrelated failing test was found... it became a nightmare to debug and though the developer and their friends tried, they could not find the bug 😭 🗿  This meant that the PR had to be redone with careful testing and often github pushing to ensure the bug did not resurface... This processed birthed PR #1475 and this collection of issues.

tl;dr;
- PR #1413 included a large number of fixes but needed to be closed due to a test failure in unrelated code which could not be fixed
- PR #1475 brought back the core functionality the PR originally meant to cover (Content Types + Fields created via YAML) but could not bring back all of the fixes from the original PR.
- This issue has been created to mark the place of one of these fixes to be brought back.

## Process:
- Look at PR #1413 and try to find the code that made this particular fix. I will try to point at some of it in this issue to aid this process.
- Create a new branch from 4.x using the branch name above
- Copy small bits of the changes from PR #1413 into this branch and commit. Only move pieces related to the fix!
  - When committing you will want to give both yourself and Stephen credit for the code! See following for how to do that (Co-authored commit).
- Push to the branch after each addition to make sure there are no test failures 😬 

### Co-Authored Commits

See [these github docs for more detail](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

For each commit including any of Stephens original code use the following. Note that you do not close the quotation after your typical commit message but rather enter two empty lines and add the co-author information.

```bash
git commit -m "Your typical commit message


Co-authored-by: Stephen Ficklin <spficklin@gmail.com>
Co-authored-by: YOUR-NAME <YOUR-NAME@EXAMPLE.COM>"
```

Example: https://github.com/tripal/tripal/pull/1475/commits/af6bc9b18b484fa7d24eda73bf2b73b6256d57cd

If needed you can alter your commit message to add this after the fact. This is easiest if done before pushing to github.

### PR DESCRIPTION
**Pull request by dsenalik**
# Tripal 4 Core Dev Task

## Description

Currently the default title for any content type is a mishmash of all the required fields. For example, analysis is `[TripalEntityType__label], [analysis_program_version], [analysis_software]`

This specifies the defaults from  the yaml file `tripal_chado/config/install/tripal.tripalentitytype_collection.default_chado.yml` - see issue #1563 for the "once upon a time" tale of this 😕

## Testing

  * Build a docker on this branch
  * Go to `/admin/structure/bio_data`
  * In the "Label" column click on analysis, i.e. `/admin/structure/bio_data/manage/bio_data_2`
  * In **Advanced Settings** you should see the title format and url format specified in the yaml file `tripal_chado/config/install/tripal.tripalentitytype_collection.default_chado.yml`

Previously, this was **[TripalEntityType__label], [analysis_program_version], [analysis_software]**, now it is:

title

Previously, this was **[type]/[TripalEntity__entity_id]**, now it is (ha ha look at that typo, let's fix it! https://github.com/tripal/tripal/pull/1583/commits/1871730397b850e4a93cddb4f6fc2075e5866ad7):

url

### CODE REVIEW

**Comment by dsenalik** [comment]
Need to add testing...

**Comment by dsenalik** [comment]
After merging in the tests from #1413
`tripal/tests/src/Functional/Services/TripalContentTypesTest.php`
Testing is failing with
```
There was 1 error:

1) Drupal\Tests\tripal\Functional\Entity\TripalContentTypesTest::testTripalContentTypes
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: You have requested a non-existent service "tripal.tripal_content_types".

/var/www/drupal9/web/core/lib/Drupal/Component/DependencyInjection/Container.php:157
/var/www/drupal9/web/core/lib/Drupal.php:197
/var/www/drupal9/web/modules/contrib/tripal/tripal/tests/src/Functional/Services/TripalContentTypesTest.php:51
/var/www/drupal9/vendor/phpunit/phpunit/src/Framework/TestResult.php:728
```

This is mentioned in only two places:
```
tripal/tests/src/Functional/Services/TripalContentTypesTest.php:51:    $content_type_service = \Drupal::service('tripal.tripal_content_types');
tripal/src/Services/TripalEntityTypeCollection.php:60:   * The YAML config file prefix is tripal.tripal_content_types.*
```

**Comment by laceysanderson** [comment]
Ahh, that particular service was renamed during #1475 to `tripal.tripalentitytype_collection` as shown here: https://github.com/tripal/tripal/blob/4.x/tripal/tripal.services.yml#L32-L36.

Also note that there was also a service named tripal_fields I believe that was also renamed: https://github.com/tripal/tripal/blob/4.x/tripal/tripal.services.yml#L27-L31

**Comment by dsenalik** [comment]
Thanks @laceysanderson, I already changed to that, but I very much appreciate the confirmation so that I can be sure 👌

**Comment by dsenalik** [comment]
Created a fresh docker on this branch, new titles are indeed being used. In this example the text displayed is just the title that I used for this analysis
Analysis

File: tripal/src/Services/TripalEntityTypeCollection.php [hunk 1]
```diff
@@ -120,36 +119,34 @@ public function validate($details) {
...
-    ## to keep track of the old bio_data_# used in previous versions of Tripal.
-    # if (array_key_exists('synonyms', $details) and !is_array($details['synonyms'])) {
-    #   $this->logger->error(t('Creation of content type, "@type", failed. The synonyms should be an array.',
-    #       ['@type' => $details['label']]));
-    #   return FALSE;
-    # }
+    if (array_key_exists('synonyms', $details) and !is_array($details['synonyms'])) {
+      $this->logger->error(t('Creation of content type, "@type", failed. The synonyms should be an array.',
+          ['@type' => $details['label']]));
+      return FALSE;
```
[the remark this answers is not in the document]
Comment (spficklin) [reply tripal/src/Services/TripalEntityTypeCollection.php:148]: Ignoring this.  We like the code the way it is.

**Comment by spficklin** [comment]
This works good!  No issues, and the code changes look good too.
